### PR TITLE
Worker restart on function timeout

### LINKED ISSUE #5391: Host process should not terminate on function timeout in OOP languages
When running with an OOP language worker, function timeouts should not cause a host process restart. In those cases, we should be able to terminate and recycle the worker without bringing the host down, addressing a class of issues related to host process termination.

### PR DESCRIPTION
**Pull request by yojagad**
Fixes #5391.

### CODE REVIEW

File: src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs [hunk 1]
```diff
@@ -405,5 +405,15 @@ public void Dispose()
             State = FunctionInvocationDispatcherState.Disposing;
             Dispose(true);
         }
+
+        public async Task RestartAsync()
+        {
+            // Dispose and restart all initialized channels
+            var channels = await GetInitializedWorkerChannelsAsync();
```
Comment (mhoeger) [inline src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs:414]: I think you should consider adding
```
State = FunctionInvocationDispatcherState.WorkerProcessRestarting;
```

Since we're re-starting all of these processes, we don't want any new requests landing on these old workers. There is still a chance that a worker process restarts and sets state back to "Initialized" (line 125) before all workers are disposed (say we have 10), but this would be a start in preventing that.

File: src/WebJobs.Script.WebHost/WebScriptHostExceptionHandler.cs [hunk 2]
```diff
@@ -48,8 +54,15 @@ public async Task OnTimeoutExceptionAsync(ExceptionDispatchInfo exceptionInfo, T
             // Task ignoreTask = _hostManager.StopAsync();
             // Give the manager and all running tasks some time to shut down gracefully.
             //await Task.Delay(timeoutGracePeriod);
-
-            _applicationLifetime.StopApplication();
+            string workerRuntime = _environment.GetEnvironmentVariable(RpcWorkerConstants.FunctionWorkerRuntimeSettingName);
+            if (string.IsNullOrEmpty(workerRuntime) || string.Equals(workerRuntime, RpcWorkerConstants.DotNetLanguageWorkerName, StringComparison.InvariantCultureIgnoreCase))
```
Comment (yojagad) [inline src/WebJobs.Script.WebHost/WebScriptHostExceptionHandler.cs]: @pragnagopa - Discussed this with @mhoeger and null could also mean out of proc in some scenarios, right? If yes, should there be a separate tracking issue where "FunctionWorkerRuntimeSettingName" should be the ultimate source of truth without ambiguities ?

Comment (pragnagopa) [reply src/WebJobs.Script.WebHost/WebScriptHostExceptionHandler.cs]: Need to rely on `FunctionInvocationDispatcherstate`.  FunctionInvocationDispatcherstate is set to initialized only for out-of-proc.

File: src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs [hunk 3]
```diff
@@ -405,5 +405,16 @@ public void Dispose()
             State = FunctionInvocationDispatcherState.Disposing;
             Dispose(true);
         }
+
+        public async Task RestartAsync()
+        {
+            // Dispose and restart all initialized channels
+            State = FunctionInvocationDispatcherState.WorkerProcessRestarting;
+            var channels = await GetInitializedWorkerChannelsAsync();
```
Comment (pragnagopa) [inline src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs:414]: You need to restart all channels not just the ones that are 'Initialized'

File: src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs (hunk 1 above)
Comment (pragnagopa) [reply src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs:414]: +1 For adding new state:WorkerProcessRestarting

File: src/WebJobs.Script.WebHost/WebScriptHostExceptionHandler.cs (hunk 2 above)
Comment (pragnagopa) [reply src/WebJobs.Script.WebHost/WebScriptHostExceptionHandler.cs]: Do not rely on 'workerRuntime' as for a given language it can change out-of-proc or not. For example, when we add a language worker for C#, this is not true.

File: src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs (hunk 3 above)
Comment (yojagad) [reply src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs:414]: What would be a scenario where a channel is in "Initializing" or "Default" state when a timeout happens? If invocations are only happening through "Initialized" channels, doesn't it make sense to restart only those? Unless I'm missing something here...

File: src/WebJobs.Script.WebHost/WebScriptHostExceptionHandler.cs (hunk 2 above)
Comment (yojagad) [reply src/WebJobs.Script.WebHost/WebScriptHostExceptionHandler.cs]: Ah, yes. This would be useless if dotnet goes out-of-proc. Agreed.

File: src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs (hunk 3 above)
Comment (pragnagopa) [reply src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs:414]: A worker channel is initialized after  worker startes and grpc channel is established...not restarting all the channels will lead to a race condition if jobhost restart is issued before grpc channel is initialized.

File: src/WebJobs.Script.WebHost/WebScriptHostExceptionHandler.cs (hunk 2 above)
Comment (mhoeger) [reply src/WebJobs.Script.WebHost/WebScriptHostExceptionHandler.cs]: What if we look for initialized channels that are actively executing the function that timed out and just shut that one down? This relies on the fact that although WebJobs [knows that the function execution has failed](https://github.com/Azure/azure-webjobs-sdk/blob/f7da27d133b9ce0aadf9918770f6ed45748f3c8a/src/Microsoft.Azure.WebJobs.Host/Executors/FunctionExecutor.cs#L291), the RpcWorkerChannel still believes that the invocation is an [executingInvocation](https://github.com/Azure/azure-functions-host/blob/7d449c87e6684cf5cbc2507ba189e3158c9d2ea6/src/WebJobs.Script/Workers/Rpc/RpcWorkerChannel.cs#L43) (maybe that's not a safe assumption in the long-run). And if nothing is found, shut down the whole application? This would help if/when we have multiple language worker processes running

File: src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs (hunk 3 above)
Comment (mhoeger) [reply src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs:414]: @pragnagopa - i think i'm also confused here. I don't think there will be a jobhost restart on a timeout? Or are you talking about a jobhost restart for a different reason?

Comment (pragnagopa) [reply src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs:414]: @mhoeger - this fix is about restarting jobhost if function invocations timeout instead of shutting down whole application

File: src/WebJobs.Script.WebHost/WebScriptHostExceptionHandler.cs [hunk 4]
```diff
@@ -48,8 +54,15 @@ public async Task OnTimeoutExceptionAsync(ExceptionDispatchInfo exceptionInfo, T
...
+            {
+                _applicationLifetime.StopApplication();
+            }
+            else
+            {
+                await _functionInvocationDispatcher.RestartAsync();
```
Comment (pragnagopa) [inline src/WebJobs.Script.WebHost/WebScriptHostExceptionHandler.cs]: Add warning level log indicating Jobhost is restarting

File: src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs (hunk 3 above)
Comment (mhoeger) [reply src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs:414]: @pragnagopa - i think the change is only to restart the worker channels, whether they are at jobhost or webhost scope (which makes sense - because functions and host configuration are not excepted to change on a timeout).

File: src/WebJobs.Script.WebHost/WebScriptHostExceptionHandler.cs (hunk 4 above)
Comment (yojagad) [reply src/WebJobs.Script.WebHost/WebScriptHostExceptionHandler.cs]: Will do. But also, this is not a jobhost restart.. Just want to make sure we're on the same page here :)

File: src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs (hunk 3 above)
Comment (yojagad) [reply src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs:414]: The change is only around restarting worker channels (and therefore recycling the processes), and not an entire jobhost restart..

Comment (pragnagopa) [reply src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs:414]: My bad. I recommend JobHost restart which aligns well with holding off reading events from event sources until jobhost restart completes. This would be cleaner approach.

File: src/WebJobs.Script.WebHost/WebScriptHostExceptionHandler.cs (hunk 2 above)
Comment (yojagad) [reply src/WebJobs.Script.WebHost/WebScriptHostExceptionHandler.cs]: @mhoeger I like that idea. Not entirely sure if this would be a much larger enhancement , hmm..

File: src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs (hunk 3 above)
Comment (pragnagopa) [reply src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs:414]: Taking my comment back. Lets test the changes where restarting channels works :)

Comment (mhoeger) [reply src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs:414]: I don't think that we wait for worker channels to be up and running to read events from event sources, so a JobHost restart seems like overkill? Although if we can change that in the future this wouldn't benefit from that change. (I ran a quick test to add a 20 second delay in sending worker init response to simulate slow startup:)
screenshot

Comment (yojagad) [reply src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs:414]: I tested this out and noticed the process id change after recycling.
image

File: src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs [hunk 5]
```diff
@@ -405,5 +405,16 @@ public void Dispose()
...
+            foreach (var channel in channels)
+            {
+                await DisposeAndRestartWorkerChannel(_workerRuntime, channel.Id);
```
Comment (mhoeger) [inline src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs:434]: Maybe too big of a change for this PR, but we should probably also drain executions here before shutting down (because a timeout, unlike an unhandled exception on the process, could have other healthy executions running too). Maybe we file an issue as a separate improvement (application shutdown has graceful timeouts on executions, so we are giving some of that up).

File: src/WebJobs.Script/Workers/Http/HttpFunctionInvocationDispatcher.cs [hunk 6]
```diff
@@ -185,5 +185,10 @@ public Task ShutdownAsync()
         {
             return Task.CompletedTask;
         }
+
+        public async Task RestartAsync()
+        {
+            await DisposeAndRestartWorkerChannel(_httpWorkerChannel.Id);
```
Comment (mhoeger) [inline src/WebJobs.Script/Workers/Http/HttpFunctionInvocationDispatcher.cs]: Add `State = FunctionInvocationDispatcherState.WorkerProcessRestarting;` here too

File: src/WebJobs.Script/Workers/Http/HttpFunctionInvocationDispatcher.cs [hunk 7]
```diff
@@ -185,5 +185,10 @@ public Task ShutdownAsync()
         {
             return Task.CompletedTask;
         }
+
+        public async Task RestartAsync()
+        {
```
Comment (mhoeger) [inline src/WebJobs.Script/Workers/Http/HttpFunctionInvocationDispatcher.cs:190]: Add `State = FunctionInvocationDispatcherState.WorkerProcessRestarting;` here too

File: src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs (hunk 5 above)
Comment (yojagad) [reply src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs:434]: We already have a grace period in the timeout exception itself though.
image

File: src/WebJobs.Script/Workers/Http/HttpFunctionInvocationDispatcher.cs (hunk 7 above)
Comment (yojagad) [reply src/WebJobs.Script/Workers/Http/HttpFunctionInvocationDispatcher.cs:190]: State is already being changed as the first step within DisposeAndRestartWorkerChannel() which is why I didn't add it here... This was not the case for RpcFunctionInvocationDispatcher though.

Comment (mhoeger) [reply src/WebJobs.Script/Workers/Http/HttpFunctionInvocationDispatcher.cs:190]: Ahh got it! I think it's in a different place for rpc because we only want to delay invocations if there are 0 workers alive. Sounds good, thanks!

File: src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs (hunk 5 above)
Comment (mhoeger) [reply src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs:434]: I think new requests are still going to be routed to the worker though? So even though there's the 2 second grace period, existing requests are finishing and new requests are just starting to execute?

Comment (yojagad) [reply src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs:434]: I think that's the behavior today as well even with normal shutdown. Even if we give grace periods for execution to complete, as long as the web app itself is accepting requests, there are chances some additional executions might flow into the worker/channel shutting down..

File: src/WebJobs.Script.WebHost/WebScriptHostExceptionHandler.cs (hunk 2 above)
Comment (pragnagopa) [reply src/WebJobs.Script.WebHost/WebScriptHostExceptionHandler.cs]: This just came up in a CRI analysis. App is using multiple worker processes - 10. Only invocation on one of them timeouted out, rest of the 9 processes are still healthy. What you and @mhoeger proposed makes more sense. It also, makes sense to only restart worker channels that are faulted due to timeout. So introduce, another state for channel 'Faulted' and only restart channels that are faulted.

File: src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs (hunk 3 above)
Comment (pragnagopa) [reply src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs:414]: This just came up in a CRI analysis. App is using multiple worker processes - 10. Only invocation on one of them timeouted out, rest of the 9 processes are still healthy. What you and @mhoeger proposed makes more sense. It also, makes sense to only restart worker channels that are faulted due to timeout. So introduce, another state for channel 'Faulted' and only restart channels that are faulted.

File: src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs (hunk 5 above)
Comment (mhoeger) [reply src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs:434]: @yojagad  - wasn't the change you made to add [DrainInvocationsAsync](https://github.com/Azure/azure-functions-host/commit/4a9aa46e9442ea9d4b0b68fbe6d5c730b8c90baf) in order to complete outstanding executions, while not allowing new ones (because the jobhost is shut down at that point)? Or maybe we need to add `State = Stopped` or something to our StopAsync to make that happen.

but either way! since we're just recycling workers we should be able to do better i think..!

(but i am actually curious if we can get concretely down the ordering of events, especially as it relates to https://github.com/Azure/azure-functions-host/issues/5785 and potentially this https://github.com/Azure/azure-functions-host/issues/5651)

**Comment by yojagad** [comment]
Changed approach for RPC channels to dispose and restart only the errored channels. @mhoeger @pragnagopa If this approach looks alright, I'll go ahead and start adding tests to it. Ran the same test as I previously did and was able to see the channel go into error state and subsequently, a different process.

**Comment by yojagad** [comment]
Changed approach to restart only channels that are in error. @mhoeger, @pragnagopa  If this approach looks alright, I'll start adding tests to it.

File: src/WebJobs.Script.WebHost/WebScriptHostExceptionHandler.cs [hunk 8]
```diff
@@ -39,17 +42,24 @@ public async Task OnTimeoutExceptionAsync(ExceptionDispatchInfo exceptionInfo, T
...
             // TODO: DI (FACAVAL) The shutdown call will invoke the host stop... but we may need to do this
             // explicitly in order to pass the timeout.
             // Task ignoreTask = _hostManager.StopAsync();
             // Give the manager and all running tasks some time to shut down gracefully.
             //await Task.Delay(timeoutGracePeriod);
-
-            _applicationLifetime.StopApplication();
+            if (_functionInvocationDispatcher.State.Equals(FunctionInvocationDispatcherState.Initialized))
+            {
+                _logger.LogWarning("Restarting language worker processes due to function timing out.", exceptionInfo.SourceException);
```
Comment (pragnagopa) [inline src/WebJobs.Script.WebHost/WebScriptHostExceptionHandler.cs]: Nit: Use similar wording `A function timeout has occurred. Restarting worker process.`

File: src/WebJobs.Script/ScriptHostBuilderExtensions.cs [hunk 9]
```diff
@@ -196,6 +195,8 @@ public static IHostBuilder AddScriptHostCore(this IHostBuilder builder, ScriptAp
                     services.TryAddEnumerable(ServiceDescriptor.Singleton<IHostedService, FunctionsScaleMonitorService>());
                 }
                 services.TryAddSingleton<FunctionsScaleManager>();
+
+                services.AddSingleton(s => s.GetRequiredService<IFunctionInvocationDispatcherFactory>().GetFunctionDispatcher());
```
Comment (pragnagopa) [inline src/WebJobs.Script/ScriptHostBuilderExtensions.cs]: Avoid adding this as a dependency. FunctionInvocationDispatcherFactory is available. Use that instead

File: src/WebJobs.Script/Workers/Rpc/RpcWorkerChannelState.cs [hunk 10]
```diff
@@ -11,13 +11,18 @@ public enum RpcWorkerChannelState
...
         /// <summary>
         /// LanguageWorkerChannel is created. Worker process is Initialized. Rpc Channel is established.
         /// </summary>
-        Initialized
+        Initialized,
+
+        /// <summary>
+        /// LanguageWorkerChannel is in error state. Worker process is Initialized. Rpc Channel is established.
+        /// </summary>
+        Errored,
```
Comment (pragnagopa) [inline src/WebJobs.Script/Workers/Rpc/RpcWorkerChannelState.cs]: Nit: Faulted?

File: src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs [hunk 11]
```diff
@@ -405,5 +405,16 @@ public void Dispose()
             State = FunctionInvocationDispatcherState.Disposing;
             Dispose(true);
         }
+
+        public async Task RestartAsync()
```
Comment (pragnagopa) [inline src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs]: Rename method. RestartFaultedChannelsAsync or RestartErroredChannelsAsync

**Review by pragnagopa — COMMENTED** [review]
Added minor comments. Approach looks good. Will do another round of reviews after unit tests are added. Thanks.

File: src/WebJobs.Script/Workers/Rpc/RpcWorkerChannelState.cs (hunk 10 above)
Comment (yojagad) [reply src/WebJobs.Script/Workers/Rpc/RpcWorkerChannelState.cs]: Does not apply anymore.

File: src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs (hunk 11 above)
Comment (yojagad) [reply src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs]: Not applicable anymore.

**Comment by yojagad** [comment]
After discussion offline, changed the approach to take advantage of using invocation id that comes with the exception to figure out which worker to restart

**Comment by mhoeger** [comment]
Draft looks good to me! This avoids any race condition between two separate timers and their callbacks.

**Comment by pragnagopa** [comment]
@yojagad @mhoeger - Earlier commit which propagated cancellation token from invoker to rpcworkerchannel is a good approach as it reduces the gap between an invocation state between Host and Rpc layer. And it will allow future extensions when we need to start sending invocation cancel messages to worker process.

You also, need to ensure to change the state for worker channel to : Faulted to ensure new invocations are not scheduled to the worker channel that is about to be restarted.

File: src/WebJobs.Script/Workers/Rpc/RpcWorkerChannelState.cs [hunk 12]
```diff
@@ -8,16 +8,16 @@ public enum RpcWorkerChannelState
...
         /// The LanguageWorkerChannel is created. Worker process is starting
         /// </summary>
-        Initializing,
+        Initializing = 1 << 1,
 
         /// <summary>
         /// LanguageWorkerChannel is created. Worker process is Initialized. Rpc Channel is established.
         /// </summary>
-        Initialized,
+        Initialized = 1 << 2,
```
Comment (pragnagopa) [inline src/WebJobs.Script/Workers/Rpc/RpcWorkerChannelState.cs]: i just sent a PR updating this to Flags :) take a look!

**Comment by yojagad** [comment]
After discussion with @Fabio and @pragnagopa , it was agreed that passing down the cancellationToken need to be taken as a separate work item. So, sticking to the current approach for now. Filed a tracking work item here https://github.com/Azure/azure-functions-host/issues/5828

Just rebased the PR and added tests. Let me know if I missed anything , please.

File: src/WebJobs.Script.WebHost/WebScriptHostExceptionHandler.cs [hunk 13]
```diff
@@ -39,17 +42,30 @@ public async Task OnTimeoutExceptionAsync(ExceptionDispatchInfo exceptionInfo, T
...
+            if (functionInvocationDispatcher.State.Equals(FunctionInvocationDispatcherState.Initialized))
+            {
+                _logger.LogWarning($"A function timeout has occurred. Restarting worker process executing invocationId '{timeoutException.InstanceId}'.", exceptionInfo.SourceException);
+                bool result = await functionInvocationDispatcher.RestartWorkerWithInvocationIdAsync(timeoutException.InstanceId.ToString());
+                if (!result)
+                {
+                    _logger.LogWarning($"Restarting all language worker processes since invocation Id '{timeoutException.InstanceId}' was not found.", exceptionInfo.SourceException);
+                    await functionInvocationDispatcher.RestartAllWorkersAsync();
+                }
+                _logger.LogWarning("Restart of language worker process(es) completed.", exceptionInfo.SourceException);
```
Comment (mhoeger) [inline src/WebJobs.Script.WebHost/WebScriptHostExceptionHandler.cs:62]: Should this be a warning or info?

Comment (yojagad) [reply src/WebJobs.Script.WebHost/WebScriptHostExceptionHandler.cs:62]: One of the previous comments suggested warning..

**Review by mhoeger — DISMISSED** [review]
✅

File: src/WebJobs.Script/Workers/Rpc/RpcWorkerChannel.cs [hunk 14]
```diff
@@ -111,7 +111,7 @@ internal class RpcWorkerChannel : IRpcWorkerChannel, IDisposable
 
         public bool IsChannelReadyForInvocations()
         {
-            return _state.HasFlag(RpcWorkerChannelState.InvocationBuffersInitialized | RpcWorkerChannelState.Initialized);
+            return !_disposing && !_disposed && _state.HasFlag(RpcWorkerChannelState.InvocationBuffersInitialized | RpcWorkerChannelState.Initialized);
```
Comment (pragnagopa) [inline src/WebJobs.Script/Workers/Rpc/RpcWorkerChannel.cs:114]: Add test where this method returns false if a restart is issued on the channel

File: src/WebJobs.Script/Workers/Rpc/RpcWorkerChannelState.cs [hunk 15]
```diff
@@ -11,21 +11,21 @@ public enum RpcWorkerChannelState
...
         /// The Default state of LanguageWorkerChannel.
         /// </summary>
-        Default = 0,
+        Default = 1 << 0,
 
         /// <summary>
         /// LanguageWorkerChannel is created. InvocationBuffers per function are setup
         /// </summary>
-        InvocationBuffersInitialized = 1,
+        InvocationBuffersInitialized = 1 << 1,
```
Comment (pragnagopa) [inline src/WebJobs.Script/Workers/Rpc/RpcWorkerChannelState.cs:19]: Any reason why this change is needed?

File: test/WebJobs.Script.Tests.Integration/Host/WebScriptHostManagerTimeoutTests.cs [hunk 16]
```diff
@@ -56,31 +56,49 @@ public async Task OnTimeoutException_UsesToken_ManagerKeepsRunning()
             }
         }
 
-        private async Task<TestFunctionHost> RunTimeoutExceptionTest(bool handleCancellation)
+        [Fact]
+        public async Task OnTimeoutException_OOP_HasExpectedLogs()
+        {
+            using (var host = await RunTimeoutExceptionTest(handleCancellation: false, argKey: "inputData", timeoutFunctionName: "TimeoutSync", path: @"TestScripts\Node"))
```
Comment (pragnagopa) [inline test/WebJobs.Script.Tests.Integration/Host/WebScriptHostManagerTimeoutTests.cs]: Also, add E2E test here: https://github.com/Azure/azure-functions-host/blob/dev/test/WebJobs.Script.Tests.Integration/WebHostEndToEnd/SamplesEndToEndTests_Node.cs ( or any language) that verifies process id is different for worker process and Hostinstance id which is same after the restart.

**Review by pragnagopa — COMMENTED** [review]
Code changes look ok.  Added more comments on adding more tests. Thanks!

File: src/WebJobs.Script/Workers/Rpc/RpcWorkerChannelState.cs (hunk 15 above)
Comment (yojagad) [reply src/WebJobs.Script/Workers/Rpc/RpcWorkerChannelState.cs:19]: I was changing the Default's value ( 0 cannot be a flag value unless it's used to represent 'None' of the flags are set ) and also updated the rest just for readability sake.

Comment (pragnagopa) [reply src/WebJobs.Script/Workers/Rpc/RpcWorkerChannelState.cs:19]: instead of using bitwise operator, for simplicity can you change the values to : 1, 2, 4 ....?

Comment (mhoeger) [reply src/WebJobs.Script/Workers/Rpc/RpcWorkerChannelState.cs:19]: I kinda prefer the bitwise operator too - it makes it easier for then next person to code and not make a math mistake!

Comment (pragnagopa) [reply src/WebJobs.Script/Workers/Rpc/RpcWorkerChannelState.cs:19]: alritie..which ever makes it easier to extend this. Will resolve this comment.

**Comment by TurtleBeach** [comment]
@pragnagopa Don't give up so easily. 4 characters (1 << 1) is more cognitive load than 1, as is the math to know what is intended. I would argue most developers do not day in / day out twiddle bits, and so the << also takes a few extra milliseconds in the brain to process.  4 characters is also 4 chances for a typo, vs 1 out of sequence integer. 1, 2, 4 are explicit. Use them. My .02

File: src/WebJobs.Script/Workers/Rpc/RpcWorkerChannel.cs (hunk 14 above)
Comment (yojagad) [reply src/WebJobs.Script/Workers/Rpc/RpcWorkerChannel.cs:114]: Added a test where this returns false when it's being disposed.

File: test/WebJobs.Script.Tests.Integration/Host/WebScriptHostManagerTimeoutTests.cs (hunk 16 above)
Comment (yojagad) [reply test/WebJobs.Script.Tests.Integration/Host/WebScriptHostManagerTimeoutTests.cs]: Added validation for node processes, how do I get the HostInstanceId from TestFunctionHost?

**Review by pragnagopa — APPROVED** [review]
👍

**Review by mhoeger — APPROVED** [review]
✅
